config/init/common/lxc-net.in

```diff
@@ -51,6 +55,52 @@ _ifup() {
     ip link set dev ${LXC_BRIDGE} up
 }
 
+start_ipv6() {
+    LXC_IPV6_ARG=""
+    if [ -n "$LXC_IPV6_ADDR" ] && [ -n "$LXC_IPV6_MASK" ] && [ -n "$LXC_IPV6_NETWORK" ]; then
+        echo 1 > /proc/sys/net/ipv6/conf/all/forwarding
+        echo 0 > /proc/sys/net/ipv6/conf/${LXC_BRIDGE}/autoconf
+        ip -6 addr add dev ${LXC_BRIDGE} ${LXC_IPV6_ADDR}/${LXC_IPV6_MASK}
+        LXC_IPV6_ARG="--dhcp-range=${LXC_IPV6_ADDR},ra-only --listen-address ${LXC_IPV6_ADDR}"
+    fi
+}
+
+start_iptables() {
+    start_ipv6
+    if [ -n "$LXC_IPV6_ARG" ] && [ "$LXC_IPV6_NAT" = "true" ]; then
+        ip6tables $use_iptables_lock -t nat -A POSTROUTING -s ${LXC_IPV6_NETWORK} ! -d ${LXC_IPV6_NETWORK} -j MASQUERADE
+    fi
+    iptables $use_iptables_lock -I INPUT -i ${LXC_BRIDGE} -p udp --dport 67 -j ACCEPT
+    iptables $use_iptables_lock -I INPUT -i ${LXC_BRIDGE} -p tcp --dport 67 -j ACCEPT
+    iptables $use_iptables_lock -I INPUT -i ${LXC_BRIDGE} -p udp --dport 53 -j ACCEPT
+    iptables $use_iptables_lock -I INPUT -i ${LXC_BRIDGE} -p tcp --dport 53 -j ACCEPT
+    iptables $use_iptables_lock -I FORWARD -i ${LXC_BRIDGE} -j ACCEPT
+    iptables $use_iptables_lock -I FORWARD -o ${LXC_BRIDGE} -j ACCEPT
+    iptables $use_iptables_lock -t nat -A POSTROUTING -s ${LXC_NETWORK} ! -d ${LXC_NETWORK} -j MASQUERADE
+    iptables $use_iptables_lock -t mangle -A POSTROUTING -o ${LXC_BRIDGE} -p udp -m udp --dport 68 -j CHECKSUM --checksum-fill
+}
+
+start_nftables() {
+    start_ipv6
+    # It is not possible to delete rules in nftables using directly the same
+    # rule that was added. Therefore, for each table we create an easily
+    # identifiable chain, for which the handler can be easily identified
+    if [ -n "$LXC_IPV6_ARG" ] && [ "$LXC_IPV6_NAT" = "true" ]; then
+        $NFT add chain ip6 nat lxc-nat
+        $NFT add rule ip6 nat lxc-nat counter masquerade
+        $NFT add rule ip6 nat postrouting ip saddr ${LXC_IPV6_NETWORK} ip daddr != ${LXC_IPV6_NETWORK} jump lxc-nat
+    fi
+    $NFT add chain inet filter lxc-filter
+    $NFT add rule inet filter lxc-filter accept
+    $NFT add rule inet filter input iifname ${LXC_BRIDGE} udp dport { 53, 67 } jump lxc-filter
+    $NFT add rule inet filter input iifname ${LXC_BRIDGE} tcp dport { 53, 67 } jump lxc-filter
+    $NFT add rule inet filter forward iifname ${LXC_BRIDGE} jump lxc-filter
```

Comment:
I'm using `ip` and `ip6` families for the `nat` table, whose hook is of type `nat`. If I understand the documentation correctly, `nat` is only supported by `ip` and `ip6`  families, but not by `inet`. 
Is there something I'm misunderstanding? https://wiki.nftables.org/wiki-nftables/index.php/Quick_reference-nftables_in_10_minutes#Chains

Reply:
The documentation does suggest that, but I'm actively using this on focal 20.04's nftables:

```
table inet tomp {
        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;

                # Outbound NAT.
                oifname $ext_if ip saddr $int_net4 snat $ext_v4
                oifname $modem_if ip saddr $int_net4 ip daddr 192.168.2.1 masquerade
        }
}
```

Reply:
Or to put it another way I've only got `table inet tomp` and inside that I've got both snat, masquerade and dnat rules working fine.

Reply:
So it might make it easier to just have one inet table called `lxc` and put everything you need inside it.

Reply:
Well, that might actually be some Ubuntu specific kernel patch, or maybe only available after a certain version. Running your script with the variables defined:
```
~$ sudo nft -c -f test-nft
test-nft:11:52-55: Error: NAT is only supported for IPv4/IPv6
                oifname $ext_if ip saddr $int_net4 snat $ext_v4
                                                   ^^^^
```

Reply:
I'm running Debian buster:
```
~$ uname -a
Linux kgpe-d16 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
```

Reply:
I think its that the kernel version of nftables is older, in LXD we only target kernel >= 5.2 due to several missing features and incompatibilities in nftables until that point. It depends how older version of nftables you're targetting.

Reply:
See https://github.com/lxc/lxd/pull/8725

Reply:
Before that we were targetting >= 5.0 for a different reason, but I forget what that was.

Reply:
Then I guess there's a bug in the nftables online documentation, which should definitely inform about that. If you think that would work, then I have no problem implementing it that way, but it should be informed to users and we should add an extra check. Otherwise adding two tables for the nat rules shouldn't be much of a hazard

Reply:
Ah, there was this comment:

https://github.com/lxc/lxd/commit/5118984d3d4b1a03e0867314d3eeb5f159387fdb#diff-390446f8c9cd49885f7ad1d2f9fe64044d1a404a4501286dbb98ba5d9d4eec58L51

"// We require a 5.x kernel to avoid weird conflicts with xtables."

Reply:
Yeah maybe check the kernel is >= 5.2, or go with separate tables for nat, which ever you feels is better for your needs. If you want to align with LXD then >= 5.2 check is the way to go.

Reply:
For me, I'd be much happier using different tables. Since LXC rules are definitely simpler than LXD rules, hopefully that does not become an issue

Reply:
Cool, in that case you may be better off just using `ip` and `ip6` `lxc` tables and then duplicating the small amount of inbound rules you need for each protocol.